Support iceberg connector concurrent insertion

liupan664021 opened the pull request:
Before this pr, if multiple clients insert into same table concurrently, the insertion can be exexuted successfully but some insertion will lost. Because the snapshot replacement required a metastore lock to make sure the isolation. Also the native `HiveTableOperations` in iceberg implements the lock guard.

```
== RELEASE NOTES ==

Iceberg Changes
* Support concurrent insertion from the same presto cluster or mutiple presto clusters which sharing the same metastore

```

v-jizhang commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -1003,6 +1012,30 @@ public synchronized void setPartitionLeases(MetastoreContext metastoreContext, S
+                Thread.sleep(10);
+            }
+            catch (InterruptedException e) {
+                lockedHiveTables.inverse().remove(hiveTableName);
```
hiveTableName was added by another thread, shouldn't be removed by this thread?

liupan664021 replied:
Thanks for review! The removal is redundant, I'll fix it.

beinan reviewed (commented):
My concern is the lock might not work properly when there are multiple presto clusters or other compute engines who are updating the certain table.  I might propose to retry the failed insert ops rather than lock the table in presto.  What do you think?

liupan664021 commented:
> My concern is the lock might not work properly when there are multiple presto clusters or other compute engines who are updating the certain table. I might propose to retry the failed insert ops rather than lock the table in presto. What do you think?

@beinan Thanks for the review! Actually there are two level lock to gurantee the `doCommit` operation: a hive metastore lock and a jvm level lock. The lock part is similar to the iceberg native [HiveTableOperations](https://github.com/apache/iceberg/blob/master/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java)

junyi1313 commented:
> > My concern is the lock might not work properly when there are multiple presto clusters or other compute engines who are updating the certain table. I might propose to retry the failed insert ops rather than lock the table in presto. What do you think?
> 
> @beinan Thanks for the review! Actually there are two level lock to gurantee the `doCommit` operation: a jvm level lock and a hive metastore lock acquired by a `ThriftHiveMetastoreClient#lock`. The lock part is similar to the iceberg native [HiveTableOperations](https://github.com/apache/iceberg/blob/master/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java)

Add a cherry pick info may be better @liupan664021

beinan commented:
> > > My concern is the lock might not work properly when there are multiple presto clusters or other compute engines who are updating the certain table. I might propose to retry the failed insert ops rather than lock the table in presto. What do you think?
> > 
> > 
> > @beinan Thanks for the review! Actually there are two level lock to gurantee the `doCommit` operation: a jvm level lock and a hive metastore lock acquired by a `ThriftHiveMetastoreClient#lock`. The lock part is similar to the iceberg native [HiveTableOperations](https://github.com/apache/iceberg/blob/master/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java)
> 
> Add a cherry pick info may be better @liupan664021


> > My concern is the lock might not work properly when there are multiple presto clusters or other compute engines who are updating the certain table. I might propose to retry the failed insert ops rather than lock the table in presto. What do you think?
> 
> @beinan Thanks for the review! Actually there are two level lock to gurantee the `doCommit` operation: a jvm level lock and a hive metastore lock acquired by a `ThriftHiveMetastoreClient#lock`. The lock part is similar to the iceberg native [HiveTableOperations](https://github.com/apache/iceberg/blob/master/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java)

Got it, that makes a lot of sense.  Thank you for the explanation!

liupan664021 commented:
@beinan @junyi1313 Sure, I'll add the cherry-pick info.

liupan664021 commented:
> My concern is the lock might not work properly when there are multiple presto clusters or other compute engines who are updating the certain table. I might propose to retry the failed insert ops rather than lock the table in presto. What do you think?

@beinan I tested concurrent insertions from two different presto cluster, and all the insertions finished without data missing.  It seems the hive metastore lock and jvm lock work well. Without the lock, even concurrent insertions from same presto cluster may cause data missing.

beinan commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -1279,6 +1290,100 @@ public void revokeTablePrivileges(MetastoreContext metastoreContext, String data
+                    .stopOnIllegalExceptions()
+                    .run("unlock",
+                        stats.getUnlock().wrap(() -> getMetastoreClientThenCall(metastoreContext, client -> {
+                            client.unlock(new UnlockRequest(lockId));
```
what if the unlock op is failed?  should we retry?

liupan664021 replied:
Same as the  lock op

beinan commented on presto-iceberg/src/main/java/com/facebook/presto/iceberg/HiveTableOperations.java:
```diff
@@ -208,72 +233,85 @@ public void commit(@Nullable TableMetadata base, TableMetadata metadata)
-        PrestoPrincipal owner = new PrestoPrincipal(USER, table.getOwner());
-        PrincipalPrivileges privileges = new PrincipalPrivileges(
-                ImmutableMultimap.<String, HivePrivilegeInfo>builder()
+            PrestoPrincipal owner = new PrestoPrincipal(USER, table.getOwner());
```
fix indentations

beinan replied:
my bad, your indentation is correct actually.

beinan commented on presto-iceberg/src/test/java/com/facebook/presto/iceberg/TestIcebergSmoke.java:
```diff
@@ -418,6 +425,47 @@ public void testInsertIntoNotNullColumn()
+        List<Thread> threads = Stream.generate(() -> new Thread(() -> {
+            int i = value.getAndIncrement();
+            try {
+                assertUpdate(session, format("INSERT INTO test_concurrent_insert VALUES(%s, '%s')", i + 1, strings[i]), 1);
```
will junit failed and collect the test result for new threads?  I think the junit only takes care of the assert in the main thread -- I might be wrong, could you confirm?

liupan664021 replied:
Thanks for review! I am not sure whether the junit assertion works in a non main thread. I only want to imitate the concurrent insertion here, but I don't know how to insert data without a assertion, so I placed a `assertUpdate` here... I'll find better way to improve it.

liupan664021 replied:
I did some change, can you have a look again?

liupan664021 commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -1279,6 +1290,100 @@ public void revokeTablePrivileges(MetastoreContext metastoreContext, String data
+            try {
+                if (state.equals(LockState.WAITING)) {
+                    retry()
+                            .maxAttempts(Integer.MAX_VALUE - 100)
```
The maxAttempts is from the Iceberg implemention. Since the acquiration of the lock is wrapped by `RetryDriver` to execute, the retries are restricted by `maxRetryTime`(30s default) and `maxAttempts`(10 default), any of which reaches the limit will abort the retries. We only want to limit the retries by the timeout, so it is reasonable to set the maxAttempts to a large count. Also, to avoid any unnecessary boundary issue, the retry is set conservatively as `Integer.MAX_VALUE - 100`.

beinan replied:
sgtm, thanks for the context!

liupan664021 commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -1279,6 +1290,100 @@ public void revokeTablePrivileges(MetastoreContext metastoreContext, String data
+                                        acquired.set(true);
+                                    }
+                                    else {
+                                        throw new RuntimeException(String.format("Failed to acquire lock: %s", newState.name()));
```
> could you help me to clarify when we will get this runtime exception?

When we acquire a lock from hive metastore, we may get the following `LockState`: `ACQUIRED`, `WAITING`, `ABORT`, `NOT_ACQUIRED`. If state is `WAITING` we can retry until timeout, but if the state is `ABORT` or `NOT_ACQUIRED`, we fail to get the lock with the runtime exception thrown.



> should we retry if we failed to acquire the lock from hive-metastore?

The acquiration of lock is wrapped by a `RetryDriver` when we call `retry().run(...)`, so there will be retries when failed to lock.

liupan664021 commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -1279,6 +1290,100 @@ public void revokeTablePrivileges(MetastoreContext metastoreContext, String data
+            }
+
+            if (!acquired.get()) {
+                throw new RuntimeException("Failed to acquire lock");
```
If the acquiration of lock timeout, in other words, retry ends without a lock acquired, we will get the runtime exception.

liupan664021 commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -1003,6 +1012,29 @@ public synchronized void setPartitionLeases(MetastoreContext metastoreContext, S
+            }
+        }
+        long lockId = currentLockId.incrementAndGet();
+        lockedHiveTables.put(lockId, hiveTableName);
```
I'll fix it.

liupan664021 commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -1003,6 +1012,29 @@ public synchronized void setPartitionLeases(MetastoreContext metastoreContext, S
+                Thread.sleep(10);
+            }
+            catch (InterruptedException e) {
+                throw new RuntimeException("Waiting for lock interrupted");
```
Thanks for suggestion! I'll fix it soon

beinan commented:
@ChunxuTang could you also help take a look?  Thanks!

liupan664021 commented:
> @ChunxuTang could you also help take a look? Thanks!

@ChunxuTang Could you help take a look of this PR? Thanks!

ChunxuTang commented:
@liupan664021 @beinan Glad to help! Will review it soon.

ChunxuTang reviewed (commented):
@liupan664021 Thanks for your nice work! 
For the commit, as Apache Iceberg and Presto are two different projects, I think it's not a cherry-pick commit, though you reuse some functionalities from there. I would suggest just linking the two Iceberg commits in your description of the PR.

ChunxuTang commented:
@beinan BTW, just for a double check: I think this feature works for the HiveMetastore mode but not the native folder mode, right?

liupan664021 commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -1003,6 +1010,30 @@ public synchronized void setPartitionLeases(MetastoreContext metastoreContext, S
+        HiveTableName hiveTableName = hiveTableName(databaseName, tableName);
+        while (lockedHiveTables.containsValue(hiveTableName)) {
+            try {
+                Thread.sleep(10);
```
Thanks so much for review! No specific reason actually. Since the `FiveHiveMetastore` only used for test, the sleep here is to mock the waiting of lock acquirement when the lock is occupied by another client or another thread more precisely. `10` is just a polling period which I think is proper.

ChunxuTang replied:
Makes sense to me. Thanks for your clarification.

liupan664021 commented:
> @liupan664021 Thanks for your nice work! For the commit, as Apache Iceberg and Presto are two different projects, I think it's not a cherry-pick commit, though you reuse some functionalities from there. I would suggest just linking the two Iceberg commits in your description of the PR.

Thanks for the suggestion! I've updated the commit message and rebased the branch.

liupan664021 commented:
@beinan @ChunxuTang Could you help have a look of this PR, thanks!

liupan664021 commented:
@beinan Could you help to merge this PR? Any comments would be appreciated~ Thanks a lot :)

beinan commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/ExtendedHiveMetastore.java:
```diff
@@ -117,4 +118,14 @@
+
+    default void unlock(MetastoreContext metastoreContext, long lockId)
+    {
+        throw new NotSupportedException("unLock is not supported by default");
```
why using an upper case 'L' in the middle?

beinan reviewed (approved):
Sorry for my delay.  it looks good to me except a couple of minor style issues.

@ChunxuTang @zhenxiao do you to wanna a second look?

zhenxiao commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -1284,6 +1295,100 @@ public void revokeTablePrivileges(MetastoreContext metastoreContext, String data
+    public long lock(MetastoreContext metastoreContext, String databaseName, String tableName)
+    {
+        try {
+            final LockComponent lockComponent = new LockComponent(LockType.EXCLUSIVE, LockLevel.TABLE, databaseName);
```
static import `LockType.EXCLUSIVE`, `LockLevel.TABLE`

liupan664021 replied:
`TABLE` has been imported by `org.apache.hadoop.hive.metastore.api.HiveObjectType.TABLE`, so I left `LockLevel.TABLE` as is.

zhenxiao commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -1284,6 +1295,100 @@ public void revokeTablePrivileges(MetastoreContext metastoreContext, String data
+            LockResponse lockResponse = stats.getLock().wrap(() -> getMetastoreClientThenCall(metastoreContext, client -> client.lock(lockRequest))).call();
+            LockState state = lockResponse.getState();
+            long lockId = lockResponse.getLockid();
+            final AtomicBoolean acquired = new AtomicBoolean(state.equals(LockState.ACQUIRED));
```
static import `LockState.ACQUIRED`

zhenxiao commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -1284,6 +1295,100 @@ public void revokeTablePrivileges(MetastoreContext metastoreContext, String data
+            final AtomicBoolean acquired = new AtomicBoolean(state.equals(LockState.ACQUIRED));
+
+            try {
+                if (state.equals(LockState.WAITING)) {
```
static import `LockState.WAITING`

zhenxiao commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -1284,6 +1295,100 @@ public void revokeTablePrivileges(MetastoreContext metastoreContext, String data
+                                getMetastoreClientThenCall(metastoreContext, client -> {
+                                    LockResponse response = client.checkLock(new CheckLockRequest(lockId));
+                                    LockState newState = response.getState();
+                                    if (newState.equals(LockState.WAITING)) {
```
static import

zhenxiao commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java:
```diff
@@ -1284,6 +1295,100 @@ public void revokeTablePrivileges(MetastoreContext metastoreContext, String data
+                                    if (newState.equals(LockState.WAITING)) {
+                                        throw new WaitingForLockException("Waiting for lock.");
+                                    }
+                                    else if (newState.equals(LockState.ACQUIRED)) {
```
static import

zhenxiao reviewed (approved):
thank you, @liupan664021 
looks good to me
could you please confirm:

- this PR only references iceberg code changes. it is not a cherry-pick from trino or other projects
- with locking, multiple Presto clusters sharing the same metatore could support concurrent insert across Presto clusters, and if so, please add documentations for it, either in commit message, or the release notes

beinan commented:
> thank you, @liupan664021 looks good to me could you please confirm:
> 
> * this PR only references iceberg code changes. it is not a cherry-pick from trino or other projects
> * with locking, multiple Presto clusters sharing the same metatore could support concurrent insert across Presto clusters, and if so, please add documentations for it, either in commit message, or the release notes

Hello @liupan664021, could you address Zhenxiao's comments?  since this pr is super useful and has been there for a while, I would like to merge it sooner rather than later.  Thanks a lot!

liupan664021 commented:
> > thank you, @liupan664021 looks good to me could you please confirm:
> > 
> > * this PR only references iceberg code changes. it is not a cherry-pick from trino or other projects
> > * with locking, multiple Presto clusters sharing the same metatore could support concurrent insert across Presto clusters, and if so, please add documentations for it, either in commit message, or the release notes
> 
> Hello @liupan664021, could you address Zhenxiao's comments? since this pr is super useful and has been there for a while, I would like to merge it sooner rather than later. Thanks a lot!

Thanks for your replies! @zhenxiao @beinan  
1. This PR only references the iceberg code changes, not a cherry-pick from any other project including Trino.
2. I tested concurrent insertions from two presto cluster which sharing the same metastore, and it works well. Also, I added the infomation to the Release notes.

abhiseksaikia commented:
@liupan664021, @beinan  This change is breaking our builds due to new methods added to HiveMetastore interface. Could you please add default implementation if that makes sense or revert this PR.

beinan commented:
@abhiseksaikia I will try to add a default implementation asap.

pratyakshsharma commented:
@beinan @liupan664021 This fix does not seem to work for more than 5 concurrent queries fired at the same time for iceberg connector. Can you please highlight how did you guys test this and if there is any configuration tuning needed to support more than 5 concurrent queries?
cc @tdcmeehan
